extensions/github1s/src/util/git-ref.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file github ref(branch or tag) utils
+ * @author netcon
+ */
+
+import * as vscode from 'vscode';
+import { reuseable } from './func';
+import { getGithubBranches, getGithubTags } from '../api';
+
+export interface RepositoryBranch {
+	name: string,
+	commit: {
+		sha: string,
+		url: string,
+	},
+	protected?: boolean
+}
+
+export interface RepositoryTag {
+	name: string,
+	commit: {
+		sha: string,
+		url:  string,
+	},
+	zipball_url: string,
+	tarball_url: string,
+	node_id: string,
+}
+
+let currentOwner = '';
+let currentRepo = '';
+let currentRef = '';
+let repositoryBranches: RepositoryBranch[] = null;
+let repositoryTags: RepositoryTag[] = null;
+
+// get current browser uri, update `currentOwner` and `currentRepo`
+const getBrowserUri = (): Promise<vscode.Uri> => {
+	return (vscode.commands.executeCommand('github1s.vscode.get-browser-url') as Promise<string>).then(browserUrl => {
+		const browserUri = vscode.Uri.parse(browserUrl);
+		const [owner = 'conwnet', repo = 'github1s'] = browserUri.path.split('/').filter(Boolean);
+		currentOwner = owner;
+		currentRepo = repo;
+		return browserUri;
+	});
+};
+
+const getRepositoryBranchesFromUri = reuseable((uri: vscode.Uri, forceUpdate: boolean = false): Promise<RepositoryBranch[]> => {
+	// use the cached branches if already fetched and not forceUpdate
+	if (repositoryBranches && repositoryBranches.length && !forceUpdate) {
+		return Promise.resolve(repositoryBranches);
+	}
+	const [owner = 'conwnet', repo = 'github1s'] = uri.path.split('/').filter(Boolean);
+	return getGithubBranches(owner, repo).then(githubBranches => (repositoryBranches = githubBranches));
+});
+
+export const getRepositoryBranches = reuseable((forceUpdate: boolean = false): Promise<RepositoryBranch[]> => {
+	return getBrowserUri().then(uri => getRepositoryBranchesFromUri(uri, forceUpdate));
+});
+
+const getRepositoryTagsFromUri = reuseable((uri: vscode.Uri, forceUpdate: boolean = false): Promise<RepositoryBranch[]> => {
+	// use the cached tags if already fetched and not forceUpdate
+	if (repositoryTags && repositoryTags.length && !forceUpdate) {
+		return Promise.resolve(repositoryTags);
+	}
+	const [owner = 'conwnet', repo = 'github1s'] = uri.path.split('/').filter(Boolean);
+	return getGithubTags(owner, repo).then(githubTags => (repositoryTags = githubTags));
+});
+
+export const getRepositoryTags = reuseable((forceUpdate: boolean = false): Promise<RepositoryBranch[]> => {
+	return getBrowserUri().then(uri => getRepositoryTagsFromUri(uri, forceUpdate));
+});
+
+
+// try to find corresponding ref from branchNames or tagNames
+const findMatchedBranchOrTag = (branchOrTagNames: string[], pathParts: string[]): string => {
+	let partIndex = 3;
+	let maybeBranch = pathParts[partIndex];
+
+	while (branchOrTagNames.find(item => item.startsWith(maybeBranch))) {
+		if (branchOrTagNames.includes(maybeBranch)) {
+			return maybeBranch;
+		}
+		maybeBranch = `${maybeBranch}/${pathParts[++partIndex]}`;
+	}
+	return null;
```

Comment:
I think there is an edge case as commit id, for example, https://github.com/conwnet/github1s/blob/9651a10f72e6471e4cd9cde852fd8f295ae6d774/package.json

We should use the `pathParts[3]` as the ref if we couldn't find possible branch or tags.